File: atlases/relatedness/shared/loaders/per_chrom_qc.js
// shared/loaders/per_chrom_qc.js
// =============================================================================
// Loads ngsPedigree Stage 2's per-chromosome QC table. One row per (dyad-or-
// triad, chromosome) with PASS/WARN/FAIL conflict counts plus the genome-wide
// trio QC fields (po_a, po_b, gw_mend_error, anc_dist, valid).
//
// Drives the right-column Mendelian-check breakdown (shared/inspector.js)
// and the Stage-1 trio_qc map used by pages/hub/inversions.js for the
// FAMILY_SUSPECT branch of the four-stage scoring.
//
// Extract-schema:
//   {
//     "schema":          "per_chrom_qc_v1",
//     "produced_by":     "ngsPedigree Stage 2",
//     "rows":            Array<row>,
//     "by_triad_chrom":  Object<"triad_id|chrom" → row>,
//     "trio_qc":         Object<triad_id → { po_a, po_b, gw_mend_error,
//                                             anc_dist, valid }>,
//   }
// =============================================================================

import { readTsv } from '../api_client.js';

export async function loadPerChromQc(args, { fetcher = null } = {}) {
  const { per_chrom_qc_path } = args || {};
  if (!per_chrom_qc_path) throw new Error('loadPerChromQc: per_chrom_qc_path is required');

  const { rows } = fetcher ? fetcher(per_chrom_qc_path) : await readTsv(per_chrom_qc_path);

  const by_triad_chrom = {};
  const trio_qc = {};
  for (const r of rows) {
    const tid = r.triad_id || r.dyad_id || r.id;
    const chr = r.chrom || r.chromosome;
    const conflicts = parseInt(r.n_conflicts || r.conflicts || '0', 10);
    const status = (r.status || '').toLowerCase();
    by_triad_chrom[`${tid}|${chr}`] = {
      triad_id: tid, chrom: chr, status, conflicts,
      raw: r,
    };
    if (!trio_qc[tid]) {
      trio_qc[tid] = {
        po_a:           r.po_a   || null,
        po_b:           r.po_b   || null,
        gw_mend_error:  parseFloat(r.gw_mend_error || '0'),
        anc_dist:       parseFloat(r.anc_dist || '0'),
        valid:          (r.valid !== 'false' && r.valid !== '0'
                          && parseFloat(r.gw_mend_error || '0') < 0.01),
      };
    }
  }

  return {
    schema:        'per_chrom_qc_v1',
    produced_by:   'ngsPedigree Stage 2',
    produced_at:   new Date().toISOString(),
    rows,
    by_triad_chrom,
    trio_qc,
  };
}

File: atlases/relatedness/registries/extractors/per_chrom_qc.py
"""extractor: per_chrom_qc_v1 — ngsPedigree Stage 2 per-chromosome QC.

Per-chromosome PASS/WARN/FAIL conflict counts per dyad/triad +
genome-wide trio QC fields. Feeds the right-column Mendelian-check
breakdown and the trio_qc map for the four-stage scoring.

Emit:
    { rows: [...], n_rows,
      pairs_summary: { n_PASS, n_WARN, n_FAIL },
      chromosomes: [chr_id, ...],
      _provenance: {...} }
"""
from __future__ import annotations
import collections
import pathlib
from typing import Any, Dict
from . import _tsv as _t


_STATUS_SYNS = ("status", "verdict", "qc_status")


def extract(raw_outputs: Dict[str, str], params: Dict[str, Any]) -> Dict[str, Any]:
    path = pathlib.Path(raw_outputs["file_path"])
    _, rows = _t.read_tsv(path, has_header=True, infer_types=True,
                          max_rows=int(params.get("max_rows") or 0))

    chroms = set()
    status_counts: Dict[str, int] = collections.Counter()
    for r in rows:
        c = r.get("chrom") or r.get("chromosome") or r.get("Chr")
        if c is not None: chroms.add(str(c))
        for k in _STATUS_SYNS:
            if k in r and r[k] is not None:
                status_counts[str(r[k])] += 1
                break

    return {
        "rows":           rows,
        "n_rows":         len(rows),
        "pairs_summary":  {f"n_{k}": v for k, v in status_counts.items()},
        "chromosomes":    sorted(chroms),
        "_provenance":    _t.provenance(raw_outputs, path, row_count=len(rows)),
    }
